### PR TITLE
Members Only

### PR DESCRIPTION
**Pull request by Rocio01**
Greetings 
This is an exclusive clubhouse app where the members can write embarrassing posts about non-members. Inside the clubhouse, members can see who the author of a post is but, outside, they can only see the story and wonder who wrote it.
Built it with Ruby on Rails using:
- Models, Views, and Controllers (MVC)
- Data Structures & Relationships
- Routing
- Migrations
- Views with forms, partials, and helpers
- RESTful design
- Adding gems for extra features

Thanks for your attention.

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -1,3 +1,7 @@
 Rails.application.routes.draw do
+  devise_for :users
+  resources :posts
```
Comment (tirthajyoti-ghosh) [inline config/routes.rb:3]: - [x] The `:posts` routes should only contain `:new`, `:create`, and  `:index` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/posts_routes_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication)).

  ```ruby
  resources :posts, only: [:new, :create, :index]
  ```

File: app/views/home/_header.html.erb [hunk 2]
```diff
@@ -0,0 +1,49 @@
+
+
+
+<nav class="navbar navbar-expand-lg navbar-dark bg-dark"> 
+
+  <div class="collapse navbar-collapse" id="navbarSupportedContent">
+    <ul class="navbar-nav mr-auto">
+    <% if user_signed_in? %>
```
Comment (tirthajyoti-ghosh) [inline app/views/home/_header.html.erb]: There should be absolutely NO logic in views. 
>Views are only for displaying information that became available to it from its corresponding controller not generating it.      

**Resources**
- [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
- [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
- [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright

- [x] Please remove all logic from all of the view files (except the ones generated by `devise`).

File: app/controllers/posts_controller.rb [hunk 3]
```diff
@@ -0,0 +1,80 @@
+class PostsController < ApplicationController
+  before_action :set_post, only: %i[show edit update destroy]
+  before_action :authenticate_user!, except: %i[index show]
+  # before_action :correct_user, only: [:edit, :update, :destroy]
```
Comment (tirthajyoti-ghosh) [inline app/controllers/posts_controller.rb]: Please remove all commented codes. If you want to access past codes, use git history.

File: app/controllers/posts_controller.rb [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  end
+
+  # GET /posts/1
+  # GET /posts/1.json
+  def show; end
+
+  # GET /posts/new
+  def new
+    @post = Post.new
+    # @post = current_user.posts.build
```
Comment (tirthajyoti-ghosh) [inline app/controllers/posts_controller.rb]: Please also remove this commented code. This is just one place where I have pointed out. Please go through your project and remove all commented codes.

File: app/models/user.rb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  has_many :friends
```
Comment (tirthajyoti-ghosh) [inline app/models/user.rb]: I think you meant `has_many :posts` here. There are no model called `friends` in your project. 
Please update this.

File: app/controllers/posts_controller.rb [hunk 6]
```diff
@@ -0,0 +1,80 @@
+class PostsController < ApplicationController
+  before_action :set_post, only: %i[show edit update destroy]
+  before_action :authenticate_user!, except: %i[index show]
```
Comment (tirthajyoti-ghosh) [inline app/controllers/posts_controller.rb:3]: You need to use a custom `before_action` method to restrict access to the `:new` and `:create` methods (the custom method should be defined in ApplicationController) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/before_action_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication)).

File: app/controllers/posts_controller.rb [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+  private
+
+  # Use callbacks to share common setup or constraints between actions.
+  def set_post
+    @post = Post.find(params[:id])
+  end
+
+  # Only allow a list of trusted parameters through.
+  def post_params
+    params.require(:post).permit(:title, :body, :author, :user_id)
```
Comment (tirthajyoti-ghosh) [inline app/controllers/posts_controller.rb]: This is not the right way to pass the `user_id` to create a new post. This poses a major security risk. Instead, use 
`current_user.posts.build(post_params)` to create a new post.

This way you are relying on rails sessions to handle user authorization, which is much more secure than passing `user_id` as params.

File: app/views/posts/_form.html.erb [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+  <div class="field">
+    <%= form.label :author %>
+    <%= form.text_field :author, class:"form-control" %>
+  </div>
+
+   <div class="field">
+    <% form.label :user_id %>
+    <%= form.number_field :user_id, id: :post_user_id, value: current_user.id, type: :hidden %>
+  </div>
```
Comment (tirthajyoti-ghosh) [inline app/views/posts/_form.html.erb]: Please remove this hidden input field. Any logged-in user can alter the `value` attribute and pose as a different author. This is a huge security flaw. 

I have already discussed a solution above. Please follow that and remove this input field.

File: app/views/posts/_form.html.erb [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+  <div class="field ">
+    <%= form.label :body %>
+    <%= form.text_area :body, class:"form-control"%>
+  </div>
+
+  <div class="field">
+    <%= form.label :author %>
+    <%= form.text_field :author, class:"form-control" %>
+  </div>
```
Comment (tirthajyoti-ghosh) [inline app/views/posts/_form.html.erb]: This field is unnecessary. You can just display the name/username of the author of the post. Asking for the author's name is unnecessary.

Please remove this field. Also, remember to remove the `author` field from the `posts` table.

**Review by tirthajyoti-ghosh — CHANGES_REQUESTED** [review]
# Changes required 🛑 

## What a remarkable project you have built 💯! This is impressive 🤩👌🧨!

### But in order to get it approved, we need to work on some changes. Let's see...

- When creating a new post, you don't need the `author` input field. You are already getting the `author` name from the logged-in user (`current_user.name`). 

I have left some additional comments below. Be sure to check them out.

Your project looks absolutely fantastic. Just make these changes and apply for another code review.

Happy coding 👨‍💻 

*Need help? Shoot me a message on Slack - [@Ghosh](https://app.slack.com/client/T47CT8XPG/DRDM6LAQN/user_profile/URB1MHXK6)*
<hr>

**Tirthajyoti Ghosh**

[GitHub](https://github.com/tirthajyoti-ghosh) | [LinkedIn](https://www.linkedin.com/in/tirthajyoti-ghosh/) | [Twitter](https://twitter.com/terrific_ghosh)

**Review by mcrd25 — APPROVED** [review]
# Changes approved :heavy_check_mark: 

Hi Team! :wave: 

Good job implementing requested changes and recognizing where you had technically alreadly implementedwhat was asked.

I see no need for further review. Maybe you can consider adding tests in the future and adding a helper method to test whether a user is signed in to be used in views.
This will make the project look even more impressive.

***If you have any questions, now or in the future, do not hesitate to contact me on social media, email or on slack and I will be glad to chat with you.***

*Happy coding 😄,*
**[--Maya Douglas](mayacdouglas.com)**
[GitHub](https://github.com/mcrd25) [Twitter](https://twitter.com/mcrd25) [linkedin](https://www.linkedin.com/in/mayadouglas/) [GitLab](https://gitlab.com/mcrd25)

**Pull request merged**